File: demo/Blackholio/client-unity/Assets/Scripts/GameManager.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Linq.Expressions;
using SpacetimeDB;
using SpacetimeDB.Types;
using UnityEngine;

public class GameManager : MonoBehaviour
{
    const string SERVER_URL = "http://127.0.0.1:3000";
    const string MODULE_NAME = "blackholio";

    public static event Action OnConnected;
    public static event Action OnSubscriptionApplied;

    public SpriteRenderer backgroundInstance;
    public float borderThickness = 2;
    public Material borderMaterial;
    public ParallaxBackground starBackgroundPrefab;
    public DeathScreen deathScreen;

    public static GameManager Instance { get; private set; }
    public static Identity LocalIdentity { get; private set; }
    public static DbConnection Conn { get; private set; }

    public static Dictionary<int, EntityController> Entities = new Dictionary<int, EntityController>();
    public static Dictionary<int, PlayerController> Players = new Dictionary<int, PlayerController>();

    private void Start()
    {
        // Clear game state in case we've disconnected and reconnected
        Entities.Clear();
        Players.Clear();
        Instance = this;
        Application.targetFrameRate = 60;

        // In order to build a connection to SpacetimeDB we need to register
        // our callbacks and specify a SpacetimeDB server URI and module name.
        var builder = DbConnection.Builder()
            .OnConnect(HandleConnect)
            .OnConnectError(HandleConnectError)
            .OnDisconnect(HandleDisconnect)
            .WithUri(SERVER_URL)
            .WithModuleName(MODULE_NAME);

        // If the user has a SpacetimeDB auth token stored in the Unity PlayerPrefs,
        // we can use it to authenticate the connection.
        // For testing purposes, it is often convenient to comment the following lines out and
        // export an executable for the project using File -> Build Settings.
        // Then, you can run the executable multiple times. Since the executable will not check for
        // a saved auth token, each run of will receive a different Identifier,
        // and their circles will be able to eat each other.
        if (AuthToken.Token != "")
        {
            Debug.Log("Using auth token!");
            builder = builder.WithToken(AuthToken.Token);
        }

        // Building the connection will establish a connection to the SpacetimeDB
        // server.
        Conn = builder.Build();
    }

    // Called when we connect to SpacetimeDB and receive our client identity
    void HandleConnect(DbConnection conn, Identity identity, string token)
    {
        Debug.Log("Connected.");
        AuthToken.SaveToken(token);
        LocalIdentity = identity;

        conn.Db.Circle.OnInsert += CircleOnInsert;
        conn.Db.Entity.OnUpdate += EntityOnUpdate;
        conn.Db.Entity.OnDelete += EntityOnDelete;
        conn.Db.Food.OnInsert += FoodOnInsert;
        conn.Db.Player.OnInsert += PlayerOnInsert;
        conn.Db.Player.OnDelete += PlayerOnDelete;

        OnConnected?.Invoke();

        // Request all tables
        Conn.SubscriptionBuilder()
            .OnApplied(HandleSubscriptionApplied)
            .SubscribeToAllTables();
    }

    void HandleConnectError(Exception ex)
    {
        Debug.LogError($"Connection error: {ex}");
    }

    void HandleDisconnect(DbConnection _conn, Exception ex)
    {
        Debug.Log("Disconnected.");
        if (ex != null)
        {
            Debug.LogException(ex);
        }
    }

    private void HandleSubscriptionApplied(SubscriptionEventContext ctx)
    {
        Debug.Log("Subscription applied!");
        OnSubscriptionApplied?.Invoke();

        // Once we have the initial subscription sync'd to the client cache
        // Get the world size from the config table and set up the arena
        var worldSize = Conn.Db.Config.Id.Find(0).WorldSize;
        SetupArena(worldSize);

        // Check to see if we already have a player, if we don't we'll need to create one
        var player = ctx.Db.Player.Identity.Find(LocalIdentity);
        if (string.IsNullOrEmpty(player.Name))
        {
            // The player has to choose a username
            UIUsernameChooser.Instance.Show(true);

            // When our username is updated, hide the username chooser
            Conn.Db.Player.OnUpdate += (_, _, newPlayer) =>
            {
                if (newPlayer.Identity == LocalIdentity)
                {
                    UIUsernameChooser.Instance.Show(false);
                }
            };
        }
        else
        {
            // We already have a player
            if (ctx.Db.Circle.PlayerId.Filter(player.PlayerId).Any())
            {
                // We already have at least one circle, we should just be able to start
                // playing immediately.
                UIUsernameChooser.Instance.Show(false);
            }
            else
            {
                // Create a new circle for our player.
                ctx.Reducers.EnterGame(player.Name);
            }
        }
    }

    private void SetupArena(float worldSize)
    {
        CreateBorderCube(new Vector2(worldSize / 2.0f, worldSize + borderThickness / 2),
            new Vector2(worldSize + borderThickness * 2.0f, borderThickness)); //North
        CreateBorderCube(new Vector2(worldSize / 2.0f, -borderThickness / 2),
            new Vector2(worldSize + borderThickness * 2.0f, borderThickness)); //South
        CreateBorderCube(new Vector2(worldSize + borderThickness / 2, worldSize / 2.0f),
            new Vector2(borderThickness, worldSize + borderThickness * 2.0f)); //East
        CreateBorderCube(new Vector2(-borderThickness / 2, worldSize / 2.0f),
            new Vector2(borderThickness, worldSize + borderThickness * 2.0f)); //West

        backgroundInstance.gameObject.SetActive(true); ;
        var size = worldSize / backgroundInstance.transform.localScale.x;
        backgroundInstance.size = new Vector2(size, size);
        backgroundInstance.transform.position = new Vector3((float)worldSize / 2, (float)worldSize / 2);

        // Set the world size for the camera controller
        CameraController.WorldSize = worldSize;
    }

    private void CreateBorderCube(Vector2 position, Vector2 scale)
    {
        var cube = GameObject.CreatePrimitive(PrimitiveType.Cube);
        cube.name = "Border";
        cube.transform.localScale = new Vector3(scale.x, scale.y, 1);
        cube.transform.position = new Vector3(position.x, position.y, 1);
        cube.GetComponent<MeshRenderer>().material = borderMaterial;
    }

    private static void CircleOnInsert(EventContext context, Circle insertedValue)
    {
        var player = GetOrCreatePlayer(insertedValue.PlayerId);
        var entityController = PrefabManager.SpawnCircle(insertedValue, player);
        Entities.Add(insertedValue.EntityId, entityController);
    }

    private static void EntityOnUpdate(EventContext context, Entity oldEntity, Entity newEntity)
    {
        if (!Entities.TryGetValue(newEntity.EntityId, out var entityController))
        {
            return;
        }
        entityController.OnEntityUpdated(newEntity);
    }

    private static void EntityOnDelete(EventContext context, Entity oldEntity)
    {
        if (Entities.Remove(oldEntity.EntityId, out var entityController))
        {
            entityController.OnDelete(context);
        }
    }

    private static void FoodOnInsert(EventContext context, Food insertedValue)
    {
        var entityController = PrefabManager.SpawnFood(insertedValue);
        Entities.Add(insertedValue.EntityId, entityController);
    }

    private static void PlayerOnInsert(EventContext context, Player insertedPlayer)
    {
        GetOrCreatePlayer(insertedPlayer.PlayerId);
    }

    private static void PlayerOnDelete(EventContext context, Player deletedvalue)
    {
        if (Players.Remove(deletedvalue.PlayerId, out var playerController))
        {
            GameObject.Destroy(playerController.gameObject);
        }
    }

    private static PlayerController GetOrCreatePlayer(int playerId)
    {
        if (!Players.TryGetValue(playerId, out var playerController))
        {
            var player = Conn.Db.Player.PlayerId.Find(playerId);
            playerController = PrefabManager.SpawnPlayer(player);
            Players.Add(playerId, playerController);
        }

        return playerController;
    }

    public static bool IsConnected()
    {
        return Conn != null && Conn.IsActive;
    }

    public void Disconnect()
    {
        Conn.Disconnect();
        Conn = null;
    }

    /* BEGIN: not in tutorial */
    private void InstanceOnUnhandledReducerError(ReducerEvent<Reducer> reducerEvent)
    {
        Debug.LogError($"There was an error!\r\n{(reducerEvent.Status as Status.Failed)?.Failed_}");
    }
    /* END: not in tutorial */
}

File: demo/Blackholio/client-ts/src/game/GameManager.ts
import type Phaser from 'phaser';
import type { Identity } from 'spacetimedb';
import { DbConnection, type ErrorContext } from '../module_bindings';
import type {
  Circle,
  ConsumeEntityEvent,
  Entity,
  Food,
  Player,
} from '../module_bindings/types';
import type { DeathScreen } from '../ui/DeathScreen';
import type { StatusHud } from '../ui/StatusHud';
import type { UsernameChooser } from '../ui/UsernameChooser';
import { CircleController } from './CircleController';
import { EntityController } from './EntityController';
import { FoodController } from './FoodController';
import { PlayerController } from './PlayerController';

const HOST = import.meta.env.VITE_SPACETIMEDB_HOST ?? 'ws://localhost:3000';
const DB_NAME = import.meta.env.VITE_SPACETIMEDB_DB_NAME ?? 'blackholio';
const TOKEN_KEY = `${HOST}/${DB_NAME}/auth_token`;

export class GameManager {
  public connection?: DbConnection;
  public identity?: Identity;
  public readonly entities = new Map<number, EntityController>();
  public readonly players = new Map<number, PlayerController>();

  private readonly pendingConsumeAnimations = new Set<number>();

  constructor(
    public readonly scene: Phaser.Scene,
    public readonly deathScreen: DeathScreen,
    private readonly usernameChooser: UsernameChooser,
    private readonly statusHud: StatusHud,
    private readonly setupArena: (worldSize: number) => void
  ) {}

  get localPlayer(): PlayerController | undefined {
    return Array.from(this.players.values()).find(player => player.isLocalPlayer);
  }

  connect(): void {
    const storedToken = localStorage.getItem(TOKEN_KEY) || undefined;
    const builder = DbConnection.builder()
      .withUri(HOST)
      .withDatabaseName(DB_NAME)
      .onConnect((connection: DbConnection, identity: Identity, token: string) => {
        this.connection = connection;
        this.identity = identity;
        localStorage.setItem(TOKEN_KEY, token);
        this.statusHud.setStatus('Connected', 'connected');
        this.registerCallbacks(connection);
        connection
          .subscriptionBuilder()
          .onApplied(() => this.handleSubscriptionApplied())
          .subscribeToAllTables();
      })
      .onDisconnect(() => {
        this.statusHud.setStatus('Disconnected', 'error');
      })
      .onConnectError((_ctx: ErrorContext, error: Error) => {
        this.statusHud.setStatus(`Error: ${error.message}`, 'error');
      });
    if (storedToken) {
      builder.withToken(storedToken);
    }
    builder.build();
  }

  enterGame(name: string): void {
    void this.connection?.reducers.enterGame({ name });
    this.usernameChooser.show(false);
    this.deathScreen.setVisible(false);
  }

  respawn(): void {
    void this.connection?.reducers.respawn({});
    this.deathScreen.setVisible(false);
  }

  findEntity(entityId: number): Entity | undefined {
    return this.connection?.db.entity.entityId.find(entityId) ?? undefined;
  }

  update(time: number, delta: number): void {
    this.entities.forEach(entity => entity.update(delta));
    this.localPlayer?.update(time);
  }

  private registerCallbacks(connection: DbConnection): void {
    connection.db.entity.onInsert((_ctx, entity) => this.syncEntity(entity.entityId));
    connection.db.entity.onUpdate((_ctx, _oldEntity, entity) => {
      this.entities.get(entity.entityId)?.onEntityUpdated(entity);
    });
    connection.db.entity.onDelete((_ctx, entity) => this.entityOnDelete(entity));
    connection.db.circle.onInsert((_ctx, circle) => this.circleOnInsert(circle));
    connection.db.circle.onDelete((_ctx, circle) => this.circleOnDelete(circle));
    connection.db.food.onInsert((_ctx, food) => this.foodOnInsert(food));
    connection.db.player.onInsert((_ctx, player) => this.playerOnInsert(player));
    connection.db.player.onUpdate((_ctx, _oldPlayer, player) =>
      this.playerOnUpdate(player)
    );
    connection.db.player.onDelete((_ctx, player) => {
      this.players.delete(player.playerId);
    });
    connection.db.consumeEntityEvent.onInsert((_ctx, event) =>
      this.consumeEntityEventOnInsert(event)
    );
  }

  private handleSubscriptionApplied(): void {
    const connection = this.connection;
    if (!connection || !this.identity) {
      return;
    }
    const config = connection.db.config.id.find(0);
    if (config) {
      this.setupArena(Number(config.worldSize));
    }
    this.syncSubscribedState();
    const player = connection.db.player.identity.find(this.identity);
    if (!player || !player.name) {
      this.usernameChooser.show(true);
      return;
    }
    const hasCircle = Array.from(connection.db.circle.iter()).some(
      circle => circle.playerId === player.playerId
    );
    if (hasCircle) {
      this.usernameChooser.show(false);
    } else {
      void connection.reducers.enterGame({ name: player.name });
    }
  }

  private syncSubscribedState(): void {
    const connection = this.connection;
    if (!connection) {
      return;
    }
    for (const player of connection.db.player.iter()) {
      this.playerOnUpdate(player);
    }
    for (const circle of connection.db.circle.iter()) {
      this.circleOnInsert(circle);
    }
    for (const food of connection.db.food.iter()) {
      this.foodOnInsert(food);
    }
  }

  private circleOnInsert(circle: Circle): void {
    if (this.entities.has(circle.entityId)) {
      return;
    }
    const entity = this.findEntity(circle.entityId);
    const player = this.getOrCreatePlayer(circle.playerId);
    if (!entity || !player) {
      return;
    }
    this.entities.set(
      circle.entityId,
      new CircleController(this.scene, entity, circle, player)
    );
  }

  private foodOnInsert(food: Food): void {
    if (this.entities.has(food.entityId)) {
      return;
    }
    const entity = this.findEntity(food.entityId);
    if (entity) {
      this.entities.set(
        food.entityId,
        new FoodController(this.scene, entity, food)
      );
    }
  }

  private circleOnDelete(circle: Circle): void {
    this.players.get(circle.playerId)?.onCircleDeleted(circle.entityId);
  }

  private syncEntity(entityId: number): void {
    const connection = this.connection;
    if (!connection || this.entities.has(entityId)) {
      return;
    }
    const circle = connection.db.circle.entityId.find(entityId);
    if (circle) {
      this.circleOnInsert(circle);
      return;
    }
    const food = connection.db.food.entityId.find(entityId);
    if (food) {
      this.foodOnInsert(food);
    }
  }

  private entityOnDelete(entity: Entity): void {
    const entityController = this.entities.get(entity.entityId);
    if (!entityController) {
      return;
    }
    this.entities.delete(entity.entityId);
    if (this.pendingConsumeAnimations.delete(entity.entityId)) {
      return;
    }
    entityController.onDelete();
  }

  private consumeEntityEventOnInsert(event: ConsumeEntityEvent): void {
    const consumedEntity = this.entities.get(event.consumedEntityId);
    const consumerEntity = this.entities.get(event.consumerEntityId);
    if (!consumedEntity || !consumerEntity) {
      return;
    }
    this.pendingConsumeAnimations.add(event.consumedEntityId);
    consumedEntity.despawnToward(consumerEntity);
  }

  private playerOnInsert(player: Player): void {
    this.getOrCreatePlayer(player.playerId);
    for (const circle of this.connection?.db.circle.iter() ?? []) {
      if (circle.playerId === player.playerId) {
        this.circleOnInsert(circle);
      }
    }
  }

  private playerOnUpdate(player: Player): void {
    const controller = this.getOrCreatePlayer(player.playerId);
    controller?.updatePlayer(player);
  }

  private getOrCreatePlayer(playerId: number): PlayerController | undefined {
    const existing = this.players.get(playerId);
    if (existing) {
      return existing;
    }
    const player = Array.from(this.connection?.db.player.iter() ?? []).find(
      value => value.playerId === playerId
    );
    if (!player) {
      return undefined;
    }
    const controller = new PlayerController(this, player);
    this.players.set(playerId, controller);
    return controller;
  }
}
